## 데이터 불러오기 및 전처리 - 서울시 대기질데이터

In [ ]:
import pandas as pd
import numpy as np

In [37]:
data_raw = pd.read_csv("data/서울시 대기질 자료 제공_2020-2021.csv",encoding="cp949")
data_raw.head()

일시   구분  미세먼지(PM10)  초미세먼지(PM25)
0  2021-05-31 23:00   평균        43.0         23.0
1  2021-05-31 23:00  강남구        42.0         24.0
2  2021-05-31 23:00  강동구        47.0         30.0
3  2021-05-31 23:00  강북구        48.0         27.0
4  2021-05-31 23:00  강서구        49.0         28.0

In [38]:
data_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 322606 entries, 0 to 322605
Data columns (total 4 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   일시           322606 non-null  object 
 1   구분           322606 non-null  object 
 2   미세먼지(PM10)   316048 non-null  float64
 3   초미세먼지(PM25)  318068 non-null  float64
dtypes: float64(2), object(2)
memory usage: 9.8+ MB


In [39]:
# 열이름 변경
data_raw.rename(columns= {'일시':'DateTime','구분':"Gu",
                         '미세먼지(PM10)':'PM10',
                         '초미세먼지(PM25)':'PM2.5'},inplace=True)

In [40]:
data_raw

DateTime   Gu  PM10  PM2.5
0       2021-05-31 23:00   평균  43.0   23.0
1       2021-05-31 23:00  강남구  42.0   24.0
2       2021-05-31 23:00  강동구  47.0   30.0
3       2021-05-31 23:00  강북구  48.0   27.0
4       2021-05-31 23:00  강서구  49.0   28.0
...                  ...  ...   ...    ...
322601   2020-01-01 0:00  용산구  12.0   13.0
322602   2020-01-01 0:00  은평구  18.0   12.0
322603   2020-01-01 0:00  종로구  20.0   15.0
322604   2020-01-01 0:00   중구  22.0   18.0
322605   2020-01-01 0:00  중랑구  22.0   10.0

[322606 rows x 4 columns]

In [41]:
data_raw[['DateTime']].dtypes

DateTime    object
dtype: object

#### 시간데이터로 변경 후 다시 데이터 타입 확인

In [50]:
# 일시에 해당하는 열 datetime으로 변경
data_raw['datetime']=pd.to_datetime(data_raw['DateTime'],format="%Y-%m-%d %H:%M") # format="%Y-%m-%d %H:%M"

In [51]:
data_raw[['datetime']].dtypes
# or
data_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 322606 entries, 0 to 322605
Data columns (total 6 columns):
 #   Column    Non-Null Count   Dtype         
---  ------    --------------   -----         
 0   DateTime  322606 non-null  object        
 1   Gu        322606 non-null  object        
 2   PM10      316048 non-null  float64       
 3   PM2.5     318068 non-null  float64       
 4   datetime  322606 non-null  datetime64[ns]
 5   yyyymmdd  322606 non-null  object        
dtypes: datetime64[ns](1), float64(2), object(3)
memory usage: 14.8+ MB


In [52]:
data_raw

DateTime   Gu  PM10  PM2.5            datetime  yyyymmdd
0       2021-05-31 23:00   평균  43.0   23.0 2021-05-31 23:00:00  20210531
1       2021-05-31 23:00  강남구  42.0   24.0 2021-05-31 23:00:00  20210531
2       2021-05-31 23:00  강동구  47.0   30.0 2021-05-31 23:00:00  20210531
3       2021-05-31 23:00  강북구  48.0   27.0 2021-05-31 23:00:00  20210531
4       2021-05-31 23:00  강서구  49.0   28.0 2021-05-31 23:00:00  20210531
...                  ...  ...   ...    ...                 ...       ...
322601   2020-01-01 0:00  용산구  12.0   13.0 2020-01-01 00:00:00  20200101
322602   2020-01-01 0:00  은평구  18.0   12.0 2020-01-01 00:00:00  20200101
322603   2020-01-01 0:00  종로구  20.0   15.0 2020-01-01 00:00:00  20200101
322604   2020-01-01 0:00   중구  22.0   18.0 2020-01-01 00:00:00  20200101
322605   2020-01-01 0:00  중랑구  22.0   10.0 2020-01-01 00:00:00  20200101

[322606 rows x 6 columns]

In [53]:
data_raw['yyyymmdd']= data_raw['datetime'].dt.strftime("%Y%m%d")

In [60]:
data_raw['hhmm'] = data_raw['datetime'].dt.strftime("%H%M")

In [61]:
data_raw.head()

DateTime   Gu  PM10  PM2.5            datetime  yyyymmdd  hhmm
0  2021-05-31 23:00   평균  43.0   23.0 2021-05-31 23:00:00  20210531  2300
1  2021-05-31 23:00  강남구  42.0   24.0 2021-05-31 23:00:00  20210531  2300
2  2021-05-31 23:00  강동구  47.0   30.0 2021-05-31 23:00:00  20210531  2300
3  2021-05-31 23:00  강북구  48.0   27.0 2021-05-31 23:00:00  20210531  2300
4  2021-05-31 23:00  강서구  49.0   28.0 2021-05-31 23:00:00  20210531  2300

In [64]:
data_raw[['hhmm']].dtypes # 문자형으로 저장됨

dtype('O')

In [66]:
data_raw=data_raw.drop('datetime',axis=1)

In [77]:
data_raw['Gu'].unique()

array(['평균', '강남구', '강동구', '강북구', '강서구', '관악구', '광진구', '구로구', '금천구',
       '노원구', '도봉구', '동대문구', '동작구', '마포구', '서대문구', '서초구', '성동구', '성북구',
       '송파구', '양천구', '영등포구', '용산구', '은평구', '종로구', '중구', '중랑구'],
      dtype=object)

In [78]:
data_raw['Gu'].value_counts()

평균      12408
강남구     12408
종로구     12408
은평구     12408
용산구     12408
영등포구    12408
양천구     12408
송파구     12408
성북구     12408
성동구     12408
서초구     12408
서대문구    12408
마포구     12408
동작구     12408
동대문구    12408
도봉구     12408
노원구     12408
금천구     12408
구로구     12408
광진구     12408
관악구     12408
강서구     12408
강북구     12408
강동구     12408
중랑구     12408
중구      12406
Name: Gu, dtype: int64

## 분석하기 - 서울시 대기질 데이터

In [67]:
data_raw_avg = data_raw.loc[data_raw['Gu']=="평균"]

In [68]:
data_raw_avg

DateTime  Gu  PM10  PM2.5  yyyymmdd  hhmm
0       2021-05-31 23:00  평균  43.0   23.0  20210531  2300
26      2021-05-31 22:00  평균  43.0   23.0  20210531  2200
52      2021-05-31 21:00  평균  45.0   25.0  20210531  2100
78      2021-05-31 20:00  평균  45.0   25.0  20210531  2000
104     2021-05-31 19:00  평균  45.0   25.0  20210531  1900
...                  ...  ..   ...    ...       ...   ...
322476   2020-01-01 4:00  평균  23.0   15.0  20200101  0400
322502   2020-01-01 3:00  평균  25.0   16.0  20200101  0300
322528   2020-01-01 2:00  평균  24.0   15.0  20200101  0200
322554   2020-01-01 1:00  평균  24.0   16.0  20200101  0100
322580   2020-01-01 0:00  평균  22.0   15.0  20200101  0000

[12408 rows x 6 columns]

In [80]:
data_raw_avg= data_raw_avg.sort_values('DateTime')
data_raw_avg

DateTime  Gu  PM10  PM2.5  yyyymmdd  hhmm
322580   2020-01-01 0:00  평균  22.0   15.0  20200101  0000
322320  2020-01-01 10:00  평균  28.0   18.0  20200101  1000
322294  2020-01-01 11:00  평균  31.0   20.0  20200101  1100
322268  2020-01-01 12:00  평균  32.0   22.0  20200101  1200
322242  2020-01-01 13:00  평균  35.0   25.0  20200101  1300
...                  ...  ..   ...    ...       ...   ...
468      2021-05-31 5:00  평균  21.0   14.0  20210531  0500
442      2021-05-31 6:00  평균  21.0   13.0  20210531  0600
416      2021-05-31 7:00  평균  22.0   14.0  20210531  0700
390      2021-05-31 8:00  평균  23.0   13.0  20210531  0800
364      2021-05-31 9:00  평균  28.0   14.0  20210531  0900

[12408 rows x 6 columns]

In [83]:
# index 다시 지정 0부터
data_raw_avg=data_raw_avg.reset_index(drop=True)

### 모든 날짜에 대한 일평균 PM10,PM2.5를 계산하세요.

In [87]:
data_raw_avg['yyyymmdd'].unique()

array(['20200101', '20200102', '20200103', '20200104', '20200105',
       '20200106', '20200107', '20200108', '20200109', '20200110',
       '20200111', '20200112', '20200113', '20200114', '20200115',
       '20200116', '20200117', '20200118', '20200119', '20200120',
       '20200121', '20200122', '20200123', '20200124', '20200125',
       '20200126', '20200127', '20200128', '20200129', '20200130',
       '20200131', '20200201', '20200202', '20200203', '20200204',
       '20200205', '20200206', '20200207', '20200208', '20200209',
       '20200210', '20200211', '20200212', '20200213', '20200214',
       '20200215', '20200216', '20200217', '20200218', '20200219',
       '20200220', '20200221', '20200222', '20200223', '20200224',
       '20200225', '20200226', '20200227', '20200228', '20200229',
       '20200301', '20200302', '20200303', '20200304', '20200305',
       '20200306', '20200307', '20200308', '20200309', '20200310',
       '20200311', '20200312', '20200313', '20200314', '202003

In [92]:
data_raw_avg_group=data_raw_avg.groupby(['yyyymmdd']).mean()

In [94]:
data_raw_avg_group.head()

PM10      PM2.5
yyyymmdd                      
20200101  32.916667  22.458333
20200102  58.375000  39.958333
20200103  67.166667  45.583333
20200104  55.916667  38.166667
20200105  51.833333  35.791667

#### 미세먼지 농도별 예보등급 구분, pm10 151이상 : 매우나쁨, 81~150 : 나쁨, 
####                                31~80 : 보통, 0~30 : 좋음


In [99]:
condition=[]

for i in data_raw_avg_group['PM10']:
    if i >= 151 :
        condition.append("VB")
    elif i>=81 :
        condition.append("B")
    elif i>=31 :
        condition.append("N")
    else:
        condition.append("G")

In [101]:
data_raw_avg_group['condition']=condition

In [102]:
data_raw_avg_group

PM10      PM2.5 condition
yyyymmdd                                
20200101  32.916667  22.458333         N
20200102  58.375000  39.958333         N
20200103  67.166667  45.583333         N
20200104  55.916667  38.166667         N
20200105  51.833333  35.791667         N
...             ...        ...       ...
20210527  40.333333  17.791667         N
20210528  39.708333  21.125000         N
20210529  20.416667  10.666667         G
20210530  24.791667  12.708333         G
20210531  33.750000  18.166667         N

[517 rows x 3 columns]

In [103]:
## PM10이 45이상인날은 ?

In [116]:
data_raw_avg_group[data_raw_avg_group['PM10']>=45]

PM10      PM2.5 condition
yyyymmdd                                 
20200102   58.375000  39.958333         N
20200103   67.166667  45.583333         N
20200104   55.916667  38.166667         N
20200105   51.833333  35.791667         N
20200109   49.541667  39.083333         N
...              ...        ...       ...
20210509   54.750000  11.875000         N
20210514   53.750000  30.250000         N
20210515   50.625000  34.500000         N
20210524  118.750000  31.250000         B
20210525  103.208333  27.625000         B

[154 rows x 3 columns]

In [117]:
sum(data_raw_avg_group['PM10']>=45) # 대기오염이45이상인날은 총 154일임

154

#### 1월달 데이터만 뽑기 

In [120]:
data_raw_avg_group['datetime']=pd.to_datetime(data_raw_avg_group.index)

In [122]:
data_raw_avg_group['yyyymm'] = data_raw_avg_group['datetime'].dt.strftime("%Y%m")
data_raw_avg_group['day'] = data_raw_avg_group['datetime'].dt.strftime("%d")

In [124]:
data_month=data_raw_avg_group.drop('datetime',axis=1)

In [126]:
data_month['yyyymm'].unique()

array(['202001', '202002', '202003', '202004', '202005', '202006',
       '202007', '202008', '202009', '202010', '202011', '202012',
       '202101', '202102', '202103', '202104', '202105'], dtype=object)

In [129]:
# 반복문을 쓰기 위한 과정
data_all=data_month['yyyymm'].unique()

In [132]:
# 초기값 설정
i=data_all[0]
data_month['yyyymm']==i

yyyymmdd
20200101     True
20200102     True
20200103     True
20200104     True
20200105     True
            ...  
20210527    False
20210528    False
20210529    False
20210530    False
20210531    False
Name: yyyymm, Length: 517, dtype: bool

In [137]:
month_temp=data_month.loc[data_month['yyyymm']==i,:]

#### csv로 저장할때  f'PM_{i}.csv' -> 특정 변수를 저장명에 넣고 싶은 경우
#### i가 가진 값 (202001)을 문자로 변형시켜줘서 이름으로 저장가능하게함.

In [139]:
month_temp.to_csv(f'PM_{i}.csv', index=False) # f'PM_{i}.csv' / index=False인덱스 저장x

In [143]:
data_raw_avg_group.head()

PM10      PM2.5 condition   datetime  yyyymm day
yyyymmdd                                                       
20200101  32.916667  22.458333         N 2020-01-01  202001  01
20200102  58.375000  39.958333         N 2020-01-02  202001  02
20200103  67.166667  45.583333         N 2020-01-03  202001  03
20200104  55.916667  38.166667         N 2020-01-04  202001  04
20200105  51.833333  35.791667         N 2020-01-05  202001  05

In [164]:
data_month[['yyyymm']].dtypes

yyyymm    object
dtype: object

In [162]:
for i in data_all : 
    month=i # str(i)
    new=data_month.loc[data_month['yyyymm']==month,:]
    new.to_csv(f'PM_{i}.csv',index=False)

In [155]:
str(i)

'202105'

# 두번째 실습 - 타슈 데이터